src/main/java/us/kbase/auth/client/internal/TokenCache.java

```diff
@@ -0,0 +1,150 @@
+	 * over the wire they're probably already strings or will be converted to
+	 * Strings at some point or another.
+	 */
+	
+	// only non-final so can be tested via reflection.
+	private static int MAX_AGE_MS = 5 * 60 * 1000; // 5 min
+	
+	final private int size;
+	final private int maxsize;
+	final private Map<String, UserDate> cache;
+	
+	/**
+	 * Create a new TokenCache.
+	 * @param size the nominal size of the cache in tokens
+	 * @param maxsize the maximum size of the cache in tokens
+	 */
+	public TokenCache(int size, int maxsize) {
+		if (size < 1 || maxsize < 1) {
+			throw new IllegalArgumentException("size and maxsize must be > 0");
+		}
+		if (size >= maxsize) {
+			throw new IllegalArgumentException("size must be < maxsize");
+		}
+		this.size = size;
+		this.maxsize = maxsize;
+		cache = new ConcurrentHashMap<String, UserDate>(maxsize);
+	}
+	
+	/** Get a token from the cache given its string. Returns null if the 
+	 * cache does not contain the token.
+	 * @param token the token string.
+	 * @return an AuthToken.
+	 */
+	public AuthToken getToken(final String token) {
+		if (token == null || token.isEmpty()) {
+			throw new IllegalArgumentException(
+					"token cannot be null or empty");
+		}
+		final UserDate ud = cache.get(getTokenDigest(token));
+		if (ud == null) {
+			return null;
+		}
+		if (new Date().getTime() - ud.date > MAX_AGE_MS) {
+			return null;
+		}
+		return new AuthToken(token, ud.user);
+	}
+	
+	private String getTokenDigest(final String token) {
+		final MessageDigest digest;
+		try {
+			digest = MessageDigest.getInstance("SHA-256");
+		} catch (NoSuchAlgorithmException e) {
+			throw new RuntimeException("Pretty sure SHA-256 is known, " +
+					"something is very broken here", e);
+		}
+		final byte[] d = digest.digest(token.getBytes(StandardCharsets.UTF_8));
+		final StringBuilder sb = new StringBuilder();
+		for (final byte b : d) {
+			sb.append(String.format("%02x", b));
+		}
+		return sb.toString();
+	}
+	
+	/**
+	 * Add a token to the cache. This method assumes the token is valid.
+	 * @param token the token to add
+	 */
+	public void putValidToken(AuthToken token) {
+		if (token == null) {
+			throw new NullPointerException("token cannot be null");
+		}
+		cache.put(getTokenDigest(token.getToken()),
+				new UserDate(token.getUserName()));
+		synchronized (cache) { // block here otherwise all threads may start sorting
+			if (cache.size() <= maxsize) {
+				return;
+			}
+			List<DateToken> dts = new ArrayList<DateToken>();
+			for (String s: cache.keySet()) {
+				dts.add(new DateToken(cache.get(s).date, s));
+			}
+			Collections.sort(dts);
+			for(int i = size; i < dts.size(); i++) {
+				cache.remove(dts.get(i).token);
+			}
+		}
+	}
+}
+
+class UserDate {
+	String user;
+	long date;
+	
+	UserDate(String user) {
+		super();
+		this.user = user;
+		this.date = new Date().getTime();
+	}
+}
+
+class DateToken implements Comparable<DateToken>{
+
+	final String token;
```

Comment:
I would rename `token` to `tokenDigest `. If my understand is correct, the key of the hashmap is definitely not token.

Reply:
This is copied verbatim from the old auth repo and is an internal only class so the ROE applies here I think, but I've made the change locally. It'll show up in a later PR

Reply:
I called it tokenHash though, that ok?

Reply:
yeah, that make sense!

Reply:
👍